ros/src/system/autoware_health_checker/src/node_status_publisher.cpp

```diff
@@ -0,0 +1,220 @@
+#include <autoware_health_checker/node_status_publisher.h>
+
+namespace autoware_health_checker
+{
+    NodeStatusPublisher::NodeStatusPublisher(ros::NodeHandle nh,ros::NodeHandle pnh)
+    {
+        node_activated_ = false;
+        nh_ = nh;
+        pnh_ = pnh;
+        status_pub_ = nh_.advertise<autoware_system_msgs::NodeStatus>("node_status",10);
+    }
+
+    NodeStatusPublisher::~NodeStatusPublisher()
+    {
+        
+    }
+
+    void NodeStatusPublisher::publishStatus()
+    {
+        ros::Rate rate = ros::Rate(autoware_health_checker::UPDATE_RATE);
+        while(ros::ok())
+        {
+            autoware_system_msgs::NodeStatus status;
+            status.node_activated = node_activated_;
+            ros::Time now = ros::Time::now();
+            status.header.stamp = now;
+            status.node_name = ros::this_node::getName();
+            std::vector<std::string> checker_keys = getRateCheckerKeys();
+            // iterate Rate checker and publish rate_check result
+            for(auto key_itr = checker_keys.begin(); key_itr != checker_keys.end(); key_itr++)
+            {
+                autoware_system_msgs::DiagnosticStatusArray diag_array;
+                autoware_system_msgs::DiagnosticStatus diag;
+                diag.type = autoware_system_msgs::DiagnosticStatus::RATE_IS_SLOW;
+                std::pair<uint8_t,double> result = rate_checkers_[*key_itr]->getErrorLevelAndRate();
+                diag.level = result.first;
+                diag.key = *key_itr;
+                diag.value = doubeToJson(result.second);
+                diag.description = rate_checkers_[*key_itr]->description;
+                diag.header.stamp = now;
+                diag_array.status.push_back(diag);
+                status.status.push_back(diag_array);
+            }
+            // iterate Diagnostic Buffer and publish all diagnostic data
+            std::vector<std::string> keys = getKeys();
+            for(auto key_itr = keys.begin(); key_itr != keys.end(); key_itr++)
+            {
+                status.status.push_back(diag_buffers_[*key_itr]->getAndClearData());
+            }
+            status_pub_.publish(status);
+            rate.sleep();
+        }
+        return;
+    }
+
+    void NodeStatusPublisher::ENABLE()
+    {
+        boost::thread publish_thread(boost::bind(&NodeStatusPublisher::publishStatus, this));
```

Comment:
Are `diag_buffers_`, `rate_checkers_`, and `node_activated_` thread-safe?

Reply:
NODE_DIACTIVATE() and NODE_ACTIVATE() is atomic operation.

Reply:
add mutex for node_status_publisher.

Reply:
rate_checker and diag_buffer Class is thread-safe.